allennlp/modules/elmo.py

```diff
@@ -496,7 +496,7 @@ def __init__(self,
                  vocab_to_cache: List[str] = None) -> None:
         super(_ElmoBiLm, self).__init__()
 
-        self._token_embedder = _ElmoCharacterEncoder(options_file, weight_file, requires_grad=requires_grad)
+        self._token_embedder = _ElmoCharacterEncoder(options_file, weight_file, requires_grad=False if vocab_to_cache is not None else requires_grad)
```

Comment:
Thanks for the fix! It'd be great if you could mention this interaction somewhere in the docstring (it's impossible to know that this happens from just looking at the docs, you have to look at the code, at this PR, or at the list of frozen parameters that AllenNLP logs).

Additionally, I'd appreciate a warning here if `vocab_to_cache is not None` and `requires_grad is True` that lets the user know that the character-level embedder will be frozen (i.e., that this new behavior will be applied).

Reply:
I think this behavior is dangerous, because even if it's documented it can have very unexpected behavior for the user.  Not everyone reads documentation (I often don't).  It's much better to raise an exception with a helpful message if `requires_grad=True` and `vocab_to_cache is not None`.

Reply:
Sure, although I don't believe I'm breaking any backward compatibility as per the bug, so the "new" behavior shouldn't be new to anyone.

Reply:
With the bug at least it didn't silently change it's behavior and instead failed loudly.  Imagine submitting a paper using ELMo where you claimed (and honestly thought) you were backproping into the the trained LM, only to find out that no, the weights weren't updated due to an obscure edge case in the framework you were using.

Reply:
@matt-peters Referring to the https://github.com/allenai/allennlp/issues/1644 can give you a better understanding of the rationale. I also gave my reasons in the PR description. If for some reason you still think it's dangerous, let me know.

Reply:
It's a good fix, and thank you for submitting the PR and adding the test.  The only change I'm suggesting is to raise an exception instead of changing the value of `requires_grad` if `vocab_to_cache` is provided.  It's about a two line change to the existing PR.

Reply:
The change only freezes the char CNN embedder (e.g. `_token_embedder`). Everything else (e.g. `_elmo_lstm`) is still parametrized by `requires_grad`. Also `_token_embedder` is not used after creating the word embedding cache.

Reply:
We do not want to raise exception. Again, if `requires_grad` is `True` and `vocab_to_cache` is provided, we still want to make `_elmo_lstm` tunable, but not so for `_token_embedder` (referring to PR description for reasons).

Reply:
Maybe the solution here is to require a third flag, where you're explicitly requesting (and thus acknowledging that you know about) the character embeddings to be fixed.  I agree with @matt-peters here that we shouldn't do this silently (and one logging statement in the midst of all of allennlp's output is pretty close to "silent" if you're not looking for the statement).

Reply:
That's an option.  The thing that worries me about this is that the `vocab_to_cache` was only intended to speed up prediction with a trained LM, and combining it with `requires_grad=True` will modify the network architecture in an implicit and non-obvious way (by effectively replacing the character CNN with a word embedding layer).

Reply:
Yeah, I agree, which is why I think we shouldn't do this without the user providing an extra flag.  That flag would be something like `freeze_cnn_layer`, or something, so you know what's happening here.  The error message when you don't pass this could say something about how this isn't really recommended, but you can do it if you want.

Reply:
I still cannot understand the concern since the character CNN layer is NOT even used after the `_ElmoBiLM.__init__` call with `vocab_to_cache`.

Reply:
@zfang the character CNN layer is used in `_ElmoBiLM.forward()` (https://github.com/allenai/allennlp/blob/cc3781dbc9f9ef56bc83b660543cdd342b7857e1/allennlp/modules/elmo.py#L578-L588), which is called in `Elmo.forward()` here (https://github.com/allenai/allennlp/blob/cc3781dbc9f9ef56bc83b660543cdd342b7857e1/allennlp/modules/elmo.py#L153) . does that clarify things a bit?

Reply:
The trouble is in what someone would assume when passing in `requires_grad=True`.  They would reasonably assume that this means, "I'm fine tuning all of the weights of the language model."  Except, if you also pass in `vocab_to_cache`, that's not true, and it's _silently_ not true.  This is a problem.

Reply:
@nelson-liu I'm talking about the `self._word_embedding is not None and word_inputs is not None` case. The block in `except RuntimeError` is a bad idea in my opinion, but that's a different story.

Reply:
@zfang backing off to using the full character embeddings is required - that cannot change. The reason for this is that we don't just want to run models we trained with the cached vocab on data we know won't be OOV - e.g for models in our demos, people might put anything in and we need to handle it smoothly.

Reply:
@DeNeutoy Hmm maybe some people are using it like that by providing character embedding ids in addition to word embedding ids. In my case (which I thought is the more usual case if you use ELMo just as a drop-in solution to replace the existing layer `torch.nn.Embedding`) I'd only provide the word embedding ids.